lib/cpp/src/thrift/transport/TSSLSocket.h

```diff
@@ -78,6 +78,7 @@ class TSSLSocket : public TSocket {
   bool peek();
   void open();
   void close();
+  bool hasPendingDataToRead();
```

Comment:
Isn't this what peek() is for?

Reply:
Peek is designed to be blocking for blocking sockets, non-blocking for non-blocking sockets. This one is non-blocking for both. I can replace the existing peek() with this implementation if you prefer that.

Reply:
It would be preferable to have one way to ask, "is there any data I can read" that does not block.

Reply:
Note that the cross test suite currently does not test asynchronous.  It is more of a protocol test; it would be interesting to add one more matrix choice of threaded vs. nonblocking for the languages that support it, and test both against both.

Reply:
@jeking3 You said it would be preferable to have one way to ask, "is there any data I can read" that does not block. I completely agree. But the existing ``peek`` method is not designed to be non-blocking. So I have a few choices here:
  * Update the doxygen comment of ``peek`` to reflect what it does and keep ``hasPendingDataToRead``
  * Add an optional argument to peek that says ``nonBlocking=false`` and if it is provided as ``true``, do what ``hasPendingDataToRead`` used to do and remove ``hasPendingDataToRead``.
  * Change ``peek`` to be always non-blocking.

How about #2 above? I think #3 is not easy, as it will require code changes in other places that I am not comfortable with.

Reply:
I implemented option #2. In terms of your matrix choice suggestion: How can we make that happen? Is that something I can help with? I am bothered by the TNonblockingSeverTest not catching the problem reported in this bug. The moment we switched to 0.11, it started throwing unexpected exceptions for us, when using Java client against C++ server. This should have been caught by the existing tests, but it did not happen. I would like to help with this. So if you have a suggestion for me, please let me know.

Reply:
I think my comments were wrong; peek blocks until there is something to do when the socket is a blocking socket.  I think that if the socket knew it was non-blocking then a call to peek() would behave like a call to hasPendingDataToRead, i.e. it would be a non-blocking call.  Perhaps that's a better way to approach it, but I would have to look at the code a little more closely.  I think the original intention of my comment was correct, there should be only one way to peek.  Calling peek() on a non-blocking socket should not block; calling peek() on a blocking socket should block.

Reply:
Sam applies to ``read``. This goes back to my original point. TSocket should behave differently for blocking vs non-blocking socket. Look at how ``TSocket`` and ``TSSLSocket`` handle ``read``s on a non-blocking socket. They throw exceptions. Luckily for me, when not using SSL, ``TNonBlockingServer`` does not ever call ``read`` on a socket that is not ready to consume data. However, that is not the case for ``SSL``.  Read calls in ``TNonBlockingServer`` are retried by catching the exception and searching for ``"retry"`` in the exception message.

The correct way to fix all this mess is to make ``TSocket`` aware of whether the socket is blocking and non-blocking and behave accordingly. Converting EAGAIN to an exception in the case of non-blocking sockets is a major performance hit. 

My suggesting is to handle the issue of making ``TSocket`` and ``TSSLSocket`` aware of the blocking/non-blocking nature of the socket in a separate Jira issue.